src/constraints/storage_constraints.jl

```diff
@@ -94,34 +96,34 @@ function add_elec_storage_dispatch_constraints(m, p, b; _n="")
     # (4h)-1: with grid
 	@constraint(m, [ts in p.time_steps_with_grid],
         m[Symbol("dvStorageEnergy"*_n)][b] >= m[Symbol("dvStoredEnergy"*_n)][b, ts-1] + p.hours_per_time_step * (  
-            sum(p.s.storage.attr[b].charge_efficiency * m[Symbol("dvProductionToStorage"*_n)][b, t, ts] for t in p.techs.elec) 
+            sum(p.s.storage.attr[b].charge_efficiency * m[Symbol("dvProductionToStorage"*_n)][b, t, ts] for t in p.techs.ac_coupled_with_storage) 
+            + sum(p.s.storage.attr[b].dc_charge_efficiency * m[Symbol("dvProductionToStorage"*_n)][b, t, ts] for t in p.techs.dc_coupled_with_storage) 
             + p.s.storage.attr[b].grid_charge_efficiency * m[Symbol("dvGridToStorage"*_n)][b, ts] 
         )
 	)
 	# (4h)-2: no grid
 	@constraint(m, [ts in p.time_steps_without_grid],
         m[Symbol("dvStorageEnergy"*_n)][b] >= m[Symbol("dvStoredEnergy"*_n)][b, ts-1] + p.hours_per_time_step * (  
-            sum(p.s.storage.attr[b].charge_efficiency * m[Symbol("dvProductionToStorage"*_n)][b,t,ts] for t in p.techs.elec) 
+            sum(p.s.storage.attr[b].charge_efficiency * m[Symbol("dvProductionToStorage"*_n)][b,t,ts] for t in p.techs.ac_coupled_with_storage) 
+            + sum(p.s.storage.attr[b].dc_charge_efficiency * m[Symbol("dvProductionToStorage"*_n)][b,t,ts] for t in p.techs.dc_coupled_with_storage) 
         )
     )
-
-	# Constraint (4i)-1: Dispatch to electrical storage is no greater than power capacity
-	@constraint(m, [ts in p.time_steps],
-        m[Symbol("dvStoragePower"*_n)][b] >= 
-            sum(m[Symbol("dvProductionToStorage"*_n)][b, t, ts] for t in p.techs.elec) + m[Symbol("dvGridToStorage"*_n)][b, ts]
-    )
 	
 	#Constraint (4k)-alt: Dispatch to and from electrical storage is no greater than power capacity
-	@constraint(m, [ts in p.time_steps_with_grid],
```

Comment:
I sometimes get confused about use of time_steps vs time_steps_with_grid and without_grid. It seems like here you have consolidated just into time_steps and removed constraint (4i)-1. Did you conclude that these were duplicative?

Reply:
Yeah I condensed three constraints ((4i)-1, (4k)-alt, (4l)-alt) into one. Constraint (4l)-alt could be removed by changing (4k)-alt to use time_steps because the dvGridToStorage term would just be zero for p.time_steps_without_grid anyway. Then constraint (4i)-1 could be removed because dvDischargeFromStorage >= 0 so (4k)-alt will always be more restrictive.